tests/test_marshalls.h

```diff
@@ -0,0 +1,329 @@
+}
+
+TEST_CASE("[Marshalls] Unsigned 64 bit integer encoding") {
+	uint8_t arr[8];
+
+	unsigned int actual_size = encode_uint64(0x0f123456789abcdef, arr);
+	CHECK(actual_size == sizeof(uint64_t));
+	CHECK_MESSAGE(arr[0] == 0xef, "First encoded byte value should be equal to low order byte value.");
+	CHECK(arr[1] == 0xcd);
+	CHECK(arr[2] == 0xab);
+	CHECK(arr[3] == 0x89);
+	CHECK(arr[4] == 0x67);
+	CHECK(arr[5] == 0x45);
+	CHECK(arr[6] == 0x23);
+	CHECK_MESSAGE(arr[7] == 0xf1, "Last encoded byte value should be equal to high order byte value.");
+}
+
+TEST_CASE("[Marshalls] Unsigned 16 bit integer decoding") {
+	uint8_t arr[] = { 0x34, 0x12 };
+
+	CHECK(decode_uint16(arr) == 0x1234);
+}
+
+TEST_CASE("[Marshalls] Unsigned 32 bit integer decoding") {
+	uint8_t arr[] = { 0x78, 0x56, 0x34, 0x12 };
+
+	CHECK(decode_uint32(arr) == 0x12345678);
+}
+
+TEST_CASE("[Marshalls] Unsigned 64 bit integer decoding") {
+	uint8_t arr[] = { 0xef, 0xcd, 0xab, 0x89, 0x67, 0x45, 0x23, 0xf1 };
+
+	CHECK(decode_uint64(arr) == 0x0f123456789abcdef);
+}
+
+TEST_CASE("[Marshalls] Floating point single precision encoding") {
+	uint8_t arr[4];
+
+	// Decimal: 0.15625
+	// IEEE 754 single-precision binary floating-point format:
+	// sign exponent (8 bits)    fraction (23 bits)
+	//  0       01111100      01000000000000000000000
+	// Hexadecimal: 0x3E200000
+	unsigned int actual_size = encode_float(0.15625f, arr);
+	CHECK(actual_size == sizeof(uint32_t));
+	CHECK(arr[0] == 0x00);
+	CHECK(arr[1] == 0x00);
+	CHECK(arr[2] == 0x20);
+	CHECK(arr[3] == 0x3e);
+}
+
+TEST_CASE("[Marshalls] Floating point double precision encoding") {
+	uint8_t arr[8];
+
+	// Decimal: 0.333333333333333314829616256247390992939472198486328125
+	// IEEE 754 double-precision binary floating-point format:
+	// sign exponent (11 bits)                  fraction (52 bits)
+	//  0      01111111101     0101010101010101010101010101010101010101010101010101
+	// Hexadecimal: 0x3FD5555555555555
+	unsigned int actual_size = encode_double(0.333333333333333314829616256247390992939472198486328125, arr);
+	CHECK(actual_size == sizeof(uint64_t));
+	CHECK(arr[0] == 0x55);
+	CHECK(arr[1] == 0x55);
+	CHECK(arr[2] == 0x55);
+	CHECK(arr[3] == 0x55);
+	CHECK(arr[4] == 0x55);
+	CHECK(arr[5] == 0x55);
+	CHECK(arr[6] == 0xd5);
+	CHECK(arr[7] == 0x3f);
+}
+
+TEST_CASE("[Marshalls] Floating point single precision decoding") {
+	uint8_t arr[] = { 0x00, 0x00, 0x20, 0x3e };
+
+	// See floating point encoding test case for details behind expected values
+	CHECK(Math::is_equal_approx(decode_float(arr), 0.15625f));
+}
+
+TEST_CASE("[Marshalls] Floating point double precision decoding") {
+	uint8_t arr[] = { 0x55, 0x55, 0x55, 0x55, 0x55, 0x55, 0xd5, 0x3f };
+
+	// See floating point encoding test case for details behind expected values
+	CHECK(Math::is_equal_approx(decode_double(arr), 0.333333333333333314829616256247390992939472198486328125));
```

Comment:
Since we are testing the validity of an encoded value, I think this should be exact equality, not approximate equality.

Also, you don't need to use the exact decimal value for the literal. Any decimal literal will automatically become the closest possible double, so we just need to put enough threes (17 digits is the safe amount) for it to [become `0.333333333333333314829616256247390992939472198486328125`](https://www.binaryconvert.com/result_double.html?decimal=048046051051051051051051051051051051051051051051051051051051))

```suggestion
	CHECK(decode_double(arr) == 0.33333333333333333));
```

Reply:
I agree, done.